File: How to write an SFTP file transfer client for WinForms/[VB]-How to write an SFTP file transfer client for WinForms/VB/ListViewItemSorters.vb
Imports System.Collections
Imports ComponentPro.Net
Imports ComponentPro.IO

Namespace ClientSample
	''' <summary>
	''' Base comparer class for comparing two list view items.
	''' </summary>
	Public MustInherit Class ListViewItemComparerBase
		Implements IComparer
		Private ReadOnly _folderImageIndex As Integer
		Private ReadOnly _folderLinkImageIndex As Integer
		Protected _sortOrder As SortOrder

		''' <summary>
		''' Initializes a new instance of the ListViewItemComparerBase.
		''' </summary>
		''' <param name="sortOrder">The sort order.</param>
		''' <param name="folderImageIndex">The defined folder image index.</param>
		''' <param name="folderLinkImageIndex">The defined folder link image index.</param>
		Protected Sub New(ByVal sortOrder As SortOrder, ByVal folderImageIndex As Integer, ByVal folderLinkImageIndex As Integer)
			_sortOrder = sortOrder

			_folderImageIndex = folderImageIndex
			_folderLinkImageIndex = folderLinkImageIndex
		End Sub

		Public Function Compare(ByVal xobject As Object, ByVal yobject As Object) As Integer Implements IComparer.Compare
			Dim x As ListViewItem = CType(xobject, ListViewItem)
			Dim y As ListViewItem = CType(yobject, ListViewItem)

			' Compare file to file or folder to folder only.
			Dim xIsFolder As Integer
			If x.ImageIndex = _folderImageIndex OrElse x.ImageIndex = _folderLinkImageIndex Then
				xIsFolder = 1
			Else
				xIsFolder = 0
			End If
			Dim yIsFolder As Integer
			If y.ImageIndex = _folderImageIndex OrElse y.ImageIndex = _folderLinkImageIndex Then
				yIsFolder = 1
			Else
				yIsFolder = 0
			End If

			If xIsFolder <> yIsFolder Then
				If _sortOrder = SortOrder.Ascending Then
					Return yIsFolder - xIsFolder
				Else
					Return (xIsFolder - yIsFolder)
				End If
			End If
			If _sortOrder = SortOrder.Ascending Then
				Return OnCompare(CType(x.Tag, ListItemInfo), CType(y.Tag, ListItemInfo), x.Text, y.Text)
			Else
				Return OnCompare(CType(y.Tag, ListItemInfo), CType(x.Tag, ListItemInfo), y.Text, x.Text)
			End If
		End Function

		Public MustOverride Function OnCompare(ByVal x As ListItemInfo, ByVal y As ListItemInfo, ByVal xtext As String, ByVal ytext As String) As Integer
	End Class

	''' <summary>
	''' Date time comparer.
	''' </summary>
	Public Class ListViewItemDateComparer
		Inherits ListViewItemComparerBase
		Public Sub New(ByVal sortOrder As SortOrder, ByVal folderImageIndex As Integer, ByVal folderLinkImageIndex As Integer)
			MyBase.New(sortOrder, folderImageIndex, folderLinkImageIndex)
		End Sub

		Public Overrides Function OnCompare(ByVal x As ListItemInfo, ByVal y As ListItemInfo, ByVal xtext As String, ByVal ytext As String) As Integer
			Return Date.Compare(x.FileInfo.LastWriteTime, y.FileInfo.LastWriteTime)
		End Function
	End Class

	''' <summary>
	''' File permissions comparer.
	''' </summary>
	Public Class ListViewItemPermissionsComparer
		Inherits ListViewItemComparerBase
		Public Sub New(ByVal sortOrder As SortOrder, ByVal folderImageIndex As Integer, ByVal folderLinkImageIndex As Integer)
			MyBase.New(sortOrder, folderImageIndex, folderLinkImageIndex)
		End Sub

		Public Overrides Function OnCompare(ByVal x As ListItemInfo, ByVal y As ListItemInfo, ByVal xtext As String, ByVal ytext As String) As Integer
			Return String.Compare(x.Permissions, x.Permissions)
		End Function
	End Class

	''' <summary>
	''' Name comparer.
	''' </summary>
	Public Class ListViewItemNameComparer
		Inherits ListViewItemComparerBase
		Public Sub New(ByVal sortOrder As SortOrder, ByVal folderImageIndex As Integer, ByVal folderLinkImageIndex As Integer)
			MyBase.New(sortOrder, folderImageIndex, folderLinkImageIndex)
		End Sub

		Public Overrides Function OnCompare(ByVal x As ListItemInfo, ByVal y As ListItemInfo, ByVal xtext As String, ByVal ytext As String) As Integer
			Return String.Compare(xtext, ytext, True)
		End Function
	End Class

	''' <summary>
	''' Size comparer.
	''' </summary>
	Public Class ListViewItemSizeComparer
		Inherits ListViewItemComparerBase
		Public Sub New(ByVal sortOrder As SortOrder, ByVal folderImageIndex As Integer, ByVal folderLinkImageIndex As Integer)
			MyBase.New(sortOrder, folderImageIndex, folderLinkImageIndex)
		End Sub

		Public Overrides Function OnCompare(ByVal x As ListItemInfo, ByVal y As ListItemInfo, ByVal xtext As String, ByVal ytext As String) As Integer
			If x.FileInfo.Length = y.FileInfo.Length Then
				Return 0
			End If
			If x.FileInfo.Length > y.FileInfo.Length Then
				Return 1
			Else
				Return -1
			End If
		End Function
	End Class
End Namespace

File: FTP Client for WinForms using Ultimate FTP library/[C#]-FTP Client for WinForms using Ultimate FTP library/C#/ListViewItemSorters.cs
using System;
using System.Collections;
using System.Windows.Forms;
using ComponentPro.Net;
using ComponentPro.IO;

namespace ClientSample
{    
    /// <summary>
    /// Base comparer class for comparing two list view items.
    /// </summary>
    public abstract class ListViewItemComparerBase : IComparer
    {
        private readonly int _folderImageIndex;
        private readonly int _folderLinkImageIndex;
        protected SortOrder _sortOrder;

        /// <summary>
        /// Initializes a new instance of the ListViewItemComparerBase.
        /// </summary>
        /// <param name="sortOrder">The sort order.</param>
        /// <param name="folderImageIndex">The defined folder image index.</param>
        /// <param name="folderLinkImageIndex">The defined folder link image index.</param>
        protected ListViewItemComparerBase(SortOrder sortOrder, int folderImageIndex, int folderLinkImageIndex)
        {
            _sortOrder = sortOrder;

            _folderImageIndex = folderImageIndex;
            _folderLinkImageIndex = folderLinkImageIndex;
        }

        public int Compare(object xobject, object yobject)
        {
            ListViewItem x = (ListViewItem) xobject;
            ListViewItem y = (ListViewItem) yobject;

            // Compare file to file or folder to folder only.
            int xIsFolder = (x.ImageIndex == _folderImageIndex || x.ImageIndex == _folderLinkImageIndex) ? 1 : 0;
            int yIsFolder = (y.ImageIndex == _folderImageIndex || y.ImageIndex == _folderLinkImageIndex) ? 1 : 0;

            if (xIsFolder != yIsFolder)
                return _sortOrder == SortOrder.Ascending ? (yIsFolder - xIsFolder) : (xIsFolder - yIsFolder);
            return _sortOrder == SortOrder.Ascending ? OnCompare((ListItemInfo)x.Tag, (ListItemInfo)y.Tag, x.Text, y.Text) : OnCompare((ListItemInfo)y.Tag, (ListItemInfo)x.Tag, y.Text, x.Text);
        }

        public abstract int OnCompare(ListItemInfo x, ListItemInfo y, string xtext, string ytext);
    }

    /// <summary>
    /// Date time comparer.
    /// </summary>
    public class ListViewItemDateComparer : ListViewItemComparerBase
    {
        public ListViewItemDateComparer(SortOrder sortOrder, int folderImageIndex, int folderLinkImageIndex) : base(sortOrder, folderImageIndex, folderLinkImageIndex)
        {
        }

        public override int OnCompare(ListItemInfo x, ListItemInfo y, string xtext, string ytext)
        {
            return DateTime.Compare(x.FileInfo.LastWriteTime, y.FileInfo.LastWriteTime);
        }
    }

    /// <summary>
    /// File permissions comparer.
    /// </summary>
    public class ListViewItemPermissionsComparer : ListViewItemComparerBase
    {
        public ListViewItemPermissionsComparer(SortOrder sortOrder, int folderImageIndex, int folderLinkImageIndex)
            : base(sortOrder, folderImageIndex, folderLinkImageIndex)
        {
        }

        public override int OnCompare(ListItemInfo x, ListItemInfo y, string xtext, string ytext)
        {
            return string.Compare(x.Permissions, x.Permissions);
        }
    }

    /// <summary>
    /// Name comparer.
    /// </summary>
    public class ListViewItemNameComparer : ListViewItemComparerBase
    {
        public ListViewItemNameComparer(SortOrder sortOrder, int folderImageIndex, int folderLinkImageIndex)
            : base(sortOrder, folderImageIndex, folderLinkImageIndex)
        {
        }

        public override int OnCompare(ListItemInfo x, ListItemInfo y, string xtext, string ytext)
        {
            return string.Compare(xtext, ytext, true);
        }
    }

    /// <summary>
    /// Size comparer.
    /// </summary>
    public class ListViewItemSizeComparer : ListViewItemComparerBase
    {
        public ListViewItemSizeComparer(SortOrder sortOrder, int folderImageIndex, int folderLinkImageIndex)
            : base(sortOrder, folderImageIndex, folderLinkImageIndex)
        {
        }

        public override int OnCompare(ListItemInfo x, ListItemInfo y, string xtext, string ytext)
        {
            if (x.FileInfo.Length == y.FileInfo.Length)
                return 0;
            return x.FileInfo.Length > y.FileInfo.Length ? 1 : -1;
        }
    }
}
